src/main/java/org/micks/DiscGolfApplication/events/DiscGolfDataService.java

```diff
@@ -56,4 +69,129 @@ public byte[] generateEventsExcel() {
             throw new RuntimeException("Failed to generate Excel file", e);
         }
     }
+    public void importEventsExcel(MultipartFile file) {
+        try (InputStream is = file.getInputStream()) {
+            Workbook workbook = new XSSFWorkbook(is);
+            Sheet sheet = workbook.getSheetAt(0);
+
+            int created = 0;
+            int updated = 0;
+
+            for (int i = 1; i <= sheet.getLastRowNum(); i++) {
+                Row row = sheet.getRow(i);
+                if (row == null) continue;
+
+                Date tournamentDate = getCellValueAsDate(row.getCell(1));
+                String pdga = getCellValueAsString(row.getCell(2));
+                String tournamentTitle = getCellValueAsString(row.getCell(3)).trim();
+                String region = getCellValueAsString(row.getCell(4));
+                String registration = getCellValueAsString(row.getCell(5));
+                String vacancies = getCellValueAsString(row.getCell(6));
+
+                if (tournamentTitle.isEmpty() || tournamentDate == null) {
+                    log.warn("Skipping row {} - missing title or date", i);
+                    continue;
+                }
+
+                DiscGolfEventDTO event = new DiscGolfEventDTO(
+                        null,
+                        tournamentDate,
+                        pdga,
+                        tournamentTitle,
+                        region,
+                        registration,
+                        vacancies
+                );
+
+                if (eventExistsByTitle(tournamentTitle)) {
+                    updateEventByTitle(tournamentTitle, event);
+                    updated++;
+                } else {
+                    discGolfEventService.createEvent(event);
+                    created++;
+                }
+            }
+
+            log.info("Import completed. Created: {}, Updated: {}", created, updated);
+
+        } catch (IOException e) {
+            throw new RuntimeException("Failed to import Excel file", e);
+        }
+    }
+
+    private String getCellValueAsString(Cell cell) {
+        if (cell == null) return "";
+
+        return switch (cell.getCellType()) {
+            case STRING -> cell.getStringCellValue();
+            case NUMERIC -> {
+                if (DateUtil.isCellDateFormatted(cell)) {
+                    yield cell.getDateCellValue().toString();
+                }
+                yield String.valueOf(cell.getNumericCellValue());
+            }
+            case BOOLEAN -> String.valueOf(cell.getBooleanCellValue());
+            default -> "";
+        };
+    }
+
+    private Date getCellValueAsDate(Cell cell) {
```

Comment:
with our own ImportException
```suggestion
    private Date getCellValueAsDate(Cell cell) {
        String dateStr = cell.getStringCellValue();
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd");
        try {
            return sdf.parse(dateStr);
        } catch (ParseException e) {
            throw new ImportException("Cannot parse date: " + dateStr);
        }
    }
```